apollo-runtime/src/main/java/com/apollographql/android/cache/normalized/RealCache.java

```diff
@@ -1,77 +1,86 @@
-import java.util.LinkedHashMap;
-import java.util.Map;
+import java.util.LinkedHashSet;
 import java.util.Set;
 import java.util.WeakHashMap;
 import java.util.concurrent.locks.ReadWriteLock;
 import java.util.concurrent.locks.ReentrantReadWriteLock;
 
 import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
 
 import static com.apollographql.android.api.graphql.util.Utils.checkNotNull;
 
 public final class RealCache implements Cache {
   private final CacheStore cacheStore;
   private final CacheKeyResolver cacheKeyResolver;
   private final ReadWriteLock lock;
-  private final Map<RecordChangeSubscriber, Set<String>> subscribers;
+  private final Set<RecordChangeSubscriber> subscribers;
 
   public RealCache(@Nonnull CacheStore cacheStore, @Nonnull CacheKeyResolver cacheKeyResolver) {
     this.cacheStore = cacheStore;
     this.cacheKeyResolver = cacheKeyResolver;
     this.lock = new ReentrantReadWriteLock();
-    this.subscribers = new WeakHashMap<>();
+    this.subscribers = Collections.newSetFromMap(new WeakHashMap<RecordChangeSubscriber, Boolean>());
   }
 
   @Override @Nonnull public ResponseNormalizer responseNormalizer() {
     return new ResponseNormalizer(cacheKeyResolver);
   }
 
-  @Override public void subscribe(RecordChangeSubscriber subscriber, Set<String> dependentKeys) {
-    subscribers.put(subscriber, dependentKeys);
+  @Override public synchronized void subscribe(RecordChangeSubscriber subscriber) {
+    subscribers.add(subscriber);
   }
 
-  @Override public void unsubscribe(RecordChangeSubscriber subscriber) {
+  @Override public synchronized void unsubscribe(RecordChangeSubscriber subscriber) {
     subscribers.remove(subscriber);
   }
 
-  @Override public void write(@Nonnull Collection<Record> recordSet) {
-    Set<String> changedKeys = Collections.emptySet();
-    lock.writeLock().lock();
-    try {
-      changedKeys = cacheStore.merge(checkNotNull(recordSet, "recordSet == null"));
-    } finally {
-      lock.writeLock().unlock();
+  @Override public void publish(@Nonnull Set<String> changedKeys) {
+    checkNotNull(changedKeys, "changedKeys == null");
+    if (changedKeys.isEmpty()) {
+      return;
+    }
+    Set<RecordChangeSubscriber> iterableSubscribers;
+    synchronized (this) {
+      iterableSubscribers = new LinkedHashSet<>(subscribers);
     }
-    Map<RecordChangeSubscriber, Set<String>> iterableSubscribers = new LinkedHashMap<>(subscribers);
-    for (Map.Entry<RecordChangeSubscriber, Set<String>> subscriberEntry : iterableSubscribers.entrySet()) {
-      if (!Utils.areDisjoint(subscriberEntry.getValue(), changedKeys)) {
-        subscriberEntry.getKey().onDependentKeysChanged();
-      }
+    for (RecordChangeSubscriber subscriber : iterableSubscribers) {
+      subscriber.onCacheKeysChanged(changedKeys);
     }
   }
 
-  @Override public Record read(@Nonnull String key) {
+  @Override public RealReadTransaction readTransaction() {
     lock.readLock().lock();
```

Comment:
Not sure this is good idea to lock in one place but unlock in `RealReadTransaction` (via closeRead).
It would be better do it in one place `RealReadTransaction` so pass lock object to `RealReadTransaction`